clang/test/Sema/builtin-counted-by-ref.c

```diff
@@ -0,0 +1,86 @@
+// RUN: %clang_cc1 -std=c99 -fsyntax-only -verify %s
+
+struct fam_struct {
+  int x;
+  char count;
+  int array[] __attribute__((counted_by(count)));
+};
+
+void test1(struct fam_struct *ptr, int size, int idx) {
+  *__builtin_counted_by_ref(ptr->array) = size;             // ok
+  *__builtin_counted_by_ref(&ptr->array[idx]) = size;       // ok
+}
+
+void test2(struct fam_struct *ptr, int idx) {
+  __builtin_counted_by_ref();                               // expected-error {{too few arguments to function call, expected 1, have 0}}
+  __builtin_counted_by_ref(ptr->array, ptr->x, ptr->count); // expected-error {{too many arguments to function call, expected 1, have 3}}
+  __builtin_counted_by_ref(ptr->x);                         // expected-error {{'__builtin_counted_by_ref' argument must reference a flexible array member}}
+  __builtin_counted_by_ref(&ptr->array[idx++]);             // expected-warning {{'__builtin_counted_by_ref' argument has side-effects that will be discarded}} expected-warning {{expression result unused}}
+}
+
+void foo(char *);
+
+void *test3(struct fam_struct *ptr, int size, int idx) {
+  char *ref = __builtin_counted_by_ref(ptr->array);         // expected-error {{value returned by '__builtin_counted_by_ref' cannot be assigned to a variable or used as a function argument}}
+
+  ref = __builtin_counted_by_ref(ptr->array);               // expected-error {{value returned by '__builtin_counted_by_ref' cannot be assigned to a variable or used as a function argument}}
+  ref = (char *)(int *)(42 + &*__builtin_counted_by_ref(ptr->array)); // expected-error {{value returned by '__builtin_counted_by_ref' cannot be assigned to a variable or used as a function argument}}
+  foo(__builtin_counted_by_ref(ptr->array));                // expected-error {{value returned by '__builtin_counted_by_ref' cannot be assigned to a variable or used as a function argument}}
```

Comment:
Worth some test coverage?
```
size_t x = sizeof(__builtin_counted_by_ref(ptr->array)); // Fine, right?
some_func_call(sizeof(__builtin_counted_by_ref(ptr->array)); // Also fine, right?

struct S {
  void *ptr;
} s;
s.ptr = __builtin_counted_by_ref(ptr->array); // Not okay, should be diagnosed

// Because the builtin seems to be available in C++, there's a lot of other fun scenarios to worry about.
auto lam = [x = __builtin_counted_by_ref(ptr->array)] { return x; }; // Not okay?

void foo(char *p = __builtin_counted_by_ref(ptr->array));
foo(); // Not okay
foo(nullptr); // Fine, right?

struct S {
  char *p = __builtin_counted_by_ref(ptr->array); // Not okay
};
// I'm guessing more examples exist as well, such as NTTP use, constant expression evaluation, etc.
```
Perhaps we don't want to expose this in C++ (yet)?

Reply:
This should be a C-only builtin (the `counted_by` attribute is C-only I believe, or at least meant to be C-only). I'll put restraints in place.

Reply:
So there doesn't seem to be a way of specifying the `Builtins.td` file that a builtin is C-only. I thought about doing something like this, but I'm not sure if it will pass muster. Thoughts?

```diff
diff --git a/clang/lib/AST/Decl.cpp b/clang/lib/AST/Decl.cpp
index b584355701b7..ead293c38910 100644
--- a/clang/lib/AST/Decl.cpp
+++ b/clang/lib/AST/Decl.cpp
@@ -3644,6 +3644,10 @@ unsigned FunctionDecl::getBuiltinID(bool ConsiderWrapperFunctions) const {
       (!hasAttr<ArmBuiltinAliasAttr>() && !hasAttr<BuiltinAliasAttr>()))
     return 0;
 
+  if (getASTContext().getLangOpts().CPlusPlus &&
+      BuiltinID == Builtin::BI__builtin_counted_by_ref)
+    return 0;
+
   const ASTContext &Context = getASTContext();
   if (!Context.BuiltinInfo.isPredefinedLibFunction(BuiltinID))
     return BuiltinID;
```

Reply:
What would be the result of this for the user? Does it emit any diagnostic or just silently ignore it?

Reply:
If `ptr->array` isn't a flexible array member, or doesn't have the `__counted_by` attribute, or the result is passed somewhere it's not supposed to go, it either becomes `(size_t *)0` or results in an error diagnostic (for the former situation).

Reply:
Note, it should probably convert to `(void *)0`, but diagnostics are generated indicating that we can't assign to type `void`.

Reply:
Oh! I think I misunderstood your question. The result is that it will treat `__builtin_counted_by_ref` as a function. From what I can tell, this is what GCC does in similar situations.